[Tang Bo Kuan] iP

bokuanT opened the pull request:
# Duke bot
> A duck a day means three ducks on wednesday
- _text-based_
- easy to use
- **never** forget your deadlines

All you need to do is:

1. download it from ~~there~~ [here](https://github.com/bokuanT/ip/releases)
2. run the jar file
3. add tasks
4. **profit**

fenced code block
```java {.line-numbers highlight-lines="1[:]"}
public class Main {
    public static void main(String[] args) {
        System.out.println("hello world");
    }
}
```

- [x] todo, deadlines and events
- [ ] contacts
- [ ] Add delight to the experience when all tasks are complete :tada:

inline code `public static void main(String[] args)`

Commit:
Add Gradle support

usuxnsnc committed:
Implemented Level 0. Greet

Added greet(), bidFarewell() methods for Duke to say hi and bye

usuxnsnc committed:
Added JavaDocs for level 0

Added docs for Duke class, greet(), bidFarewell() and makeSeparation() methods

usuxnsnc committed:
Finalise increment 0

usuxnsnc committed:
Completed implementation of Level 1. Greet, Echo, Exit

Added takeCommand() and makeResponse() methods.

usuxnsnc committed:
Completed Level 2. Add, List

Duke now can store tasks
added array to store task
modified makeResponse() to handle 'list' command
added '\t' escape sequence to Duke's responses

usuxnsnc committed:
Completed implementation of Level 3. Mark as Done

Added Task class
Added checkForCommand() method to check for commands. Improved command handling in makeResponse() method

usuxnsnc committed:
Completed implementation of Level 4. ToDOs, Events, Deadlines

Added ToDo, Event and Deadline classes, subclasses of Task class
Added CreareTask() method to handle the creation of correct Task subclasses
Implemented basic error handling using print statements
Added semi automated test-cases, to be tested using runtest.bat

usuxnsnc committed:
Completed Level 5. Handle Errors

Added DukeException, DukeUnknownInputException and DukeBadInstructionFormatException
Duke now handles wrong commands and wrong format for tasks

usuxnsnc committed:
Completed Level 6. Delete

Added delete function to Duke.
Added DukeMemory class, which acts as a task manager.
Added enum for commands like "mark", "delete" etc.
Updated Automated Testing.

usuxnsnc committed:
Add small changes in Duke.java

If-else statements are harder to read than switch statements. There
are no javadocs for Commands enum.

Better coding practices from CS2103T can be followed.

Using switch statements make code more readable than if-else.

Let's convert if-else statements to switch statements when possible.

Switch statements are easier to read than if-else statements.

usuxnsnc committed:
Add save feature to Duke

Level 6 Duke does not have 'save' feature. Level 7 increment
requires 'save' feature.

Let's.
* add FileManager class to keep track of duke.txt, the 'hard disk'
* update automated testing runtest.bat and input to support testing
this new feature

usuxnsnc committed:
Add feature to understand date and times

Duke uses String to store 'by' of deadlines or 'from' and 'to' of
events.

This is not utilizing available package LocalDateTime to display
dates and times better.

Let's,
* Use LocalDateTime for subclasses of Task that need to store a
date and time instead of Strings.
* Display date time in 'dd LLL yyyy HHmm' format.
* Get user date time input in 'dd/MM/yyyy HHmm' format.

usuxnsnc committed:
Merge branch 'branch-Level-7'

usuxnsnc committed:
Merge branch 'branch-Level-8'

usuxnsnc committed:
Merge Save feature with date and time feature

Loading Tasks from memory: Deadline and Event date time format in
duke.txt not compatible with constructor parameter format.

Stored date time format is compatible with constructor date time
format.

Let's,
* Add variable to store acceptable date time format strings
* Update automated testing scripts

usuxnsnc committed:
Improve OOP usage and organization

All classes are in the same directory and the main duke file is very
long as a lot of logic that can be abstracted to other classes are
in there.

Better OOP practices need to be followed and the main duke class
needs to be more readable and less cluttered.

Lets,
* Encapsulate logic behind parsing user input and printing to output
into the Ui and Parser classes
* Encapsulate all commands using Command classes and move away from
simple enumerations
* Package and refactor files into better organized directories.
* Update the automated testing scripts

usuxnsnc committed:
Merge branch 'add-gradle-support' of https://github.com/bokuanT/ip

usuxnsnc committed:
Integrate gradle, JUnit and package duke as a JAR file

We do not have build automation or JUnit testing.

Week 3 iP task 4 requires the addition of gradle, Junit testing and
having a duke.jar.

Let's,
* Merge the gradle fork available
* Add simple JUnit tests for TaskList and Deadline
* Create a Jar file (make it available on new release page)

usuxnsnc committed:
Fix style errors and add javadocs to main classes

Not all classes have javadocs or match the code style.

We know its good practice to do so.

Let's,
* Add javadocs to all non-test classes
* Correct the wrong style formats

usuxnsnc committed:
Finalize coding standard conventions

Style conventions are not fully adhered to.

It is good practice to follow the given coding standard.

Let's,
* Strictly follow the coding standard and pass gradle's checkstyles

usuxnsnc committed:
Add find function

Duke does not have a 'find' function.

Level 9 requires us to add it.

Let's,
* Add a 'find' function to Duke
* Add a FindCommand class to encapsulate this

usuxnsnc committed:
Merge branch 'branch-A-CodingStandard1' into branch-A-CodingStandard

usuxnsnc committed:
Merge branch 'branch-A-CodingStandard'

usuxnsnc committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/tasklist/TaskList.java

usuxnsnc committed:
Implement rough but functional Duke with GUI

Let's,
* Integrate javafx for a GUI.
* Add relevant classes like DialogBox.
* Make Duke.java compatible with javafx launcher
* Make Duke the entrypoint of the javafx launcher

usuxnsnc committed:
Implement rough but functional Duke with GUI

Let's,
* Integrate javafx for a GUI.
* Add relevant classes like DialogBox.
* Make Duke.java compatible with javafx launcher
* Make Duke the entrypoint of the javafx launcher

(cherry picked from commit 0215319e4d6a7c11f31f5f852968c0bb6f8d9d4a)

usuxnsnc committed:
Make temporary commit

jinnieshin commented on data/tasks.txt:
```diff
@@ -0,0 +1,2 @@
+E@false@book fair@11/01/2000 1000@12/01/2000 1200
+D@false@task a@11/11/2000 1900
```
Would you add this file to .gitignore so that the user data would not be revealed?

bokuanT replied:
Thank you for catching that! It would indeed be bad if the user’s data is leaked

jinnieshin commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,77 @@
+     * Starts Duke's backend processes.
+     */
+    public void run() {
+        ui.showInitMessage();
+        ui.showGreeting();
```
Is there a reason why you separated the InitMessage and Greeting into separate functions?

bokuanT replied:
I didn’t have a particular reason, but after the addition of GUI, I may not want the init message to be said by duke anymore as it isn’t really ‘speakable’

jinnieshin commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+        return "[D]" + super.toString() + " (by: " + Task.getDateTimeString(this.by) + ")";
+    }
+
+    @Override
+    public String getFileFormatString() {
+        //to be split using "|"
+        return "D" + "@" + this.isDone + "@" + this.description + "@" + this.byString;
+    }
```
I am not sure about what do you mean by 'to be split using "|"' as there is no "|" in your code. Could you please explain? I noticed the same issue in Event and ToDo classes too.

bokuanT replied:
This is a mistake on my part for not updating the comments

jinnieshin commented on text-ui-test/ACTUAL2.TXT:
```diff
@@ -0,0 +1,21 @@
+Existing tasks found. Loaded existing tasks.
+Hello from
+ ____        _        
+|  _ \ _   _| | _____ 
+| | | | | | | |/ / _ \
+| |_| | |_| |   <  __/
+|____/ \__,_|_|\_\___|
+
+	Hello! I'm Duke
+	What can I do for you?
+	____________________________________________________________
+	____________________________________________________________
+	1. [T][X] read book
+	2. [D][ ] return book (by: 06 Jun 2023 1800)
+	3. [T][X] join sports club
+	4. [T][ ] borrow book
+	5. [E][ ] project meeting (from: 22 Jan 2023 0800 to: 22 Jan 2023 2200)
+	____________________________________________________________
+	____________________________________________________________
+	Bye. Hope to see you again soon!
+	____________________________________________________________
```
I like that you have been conducting TextUiTest continuously👍 I should also do this.

jinnieshin reviewed (commented):
Looks good. I like how you took care of small details!

usuxnsnc committed:
Merge branch 'branch-Level-10-rough' into branch-Level-10

usuxnsnc committed:
Add GUI to Duke and fix minor issues

Duke is only works on CLI, minimizing user satisfaction.

Duke may seem unappealing to some users.

Let's,
* Integrate Javafx to give Duke a GUI
* Use fxml files to make code follow OOP principles better
* Update minor details like comments based on PR review feedback

usuxnsnc committed:
Merge branch 'branch-Level-10'

usuxnsnc committed:
Add assert to check certain commands and fix 'bye' bug

We assume the full command inputted by the user is correct when
instantiating commands like todo, event and deadline. 'bye' also
does not close Duke.

We should explicitly check the invariant that the full command
corresponds to the correct command type.

Let's,
*Add assert checks for todo, deadline, event and bye commands
*Make the GUI close upon the 'bye' command

bokuanT committed:
Merge pull request #1 from bokuanT/branch-A-Assertions

Add assert to check certain commands and fix 'bye' bug

usuxnsnc committed:
Improve code quality

Maintaining good code quality is important throughout development.

Certain new features from the previous few pushes do not follow good
code quality.

Let's,
*Reduce copy pasting in Ui and only declare fixed messages once,
avoiding 'magic' strings
*SLAP harder in task commands by abstracting the process of getting
task descriptions, event from, event to and deadline by in the
command class

bokuanT committed:
Merge pull request #2 from bokuanT/branch-A-CodeQuality

Improve code quality

usuxnsnc committed:
Merge branch 'master' of https://github.com/bokuanT/ip

usuxnsnc committed:
Merge branch 'master' into branch-A-CodeQuality

usuxnsnc committed:
Merge branch 'master' into branch-A-Assertions

usuxnsnc committed:
Add CI using github Actions

usuxnsnc committed:
Add remind feature, abstract LocalDateTime methods, improve code
quality

Duke cannot remind a user if a task is due on the current day. There
are also magic symbols to split storage text and a lot of methods
using LocalDateTime is in the Task class.

We want to add a new feature called remind and we should encapsulate
LocalDateTime in a more relevant class.

Let's,
*Add remind command that reminds user on events and deadlines due
today
*Refactor LocalDateTime methods to a new Time class
*Make minor changes to improve code quality like removing magic
symbols

usuxnsnc committed:
Revert "Add CI using github Actions"

This reverts commit 260b062a36deef1d5a67570f331cf60fd313d45a.

usuxnsnc committed:
Revert "Revert "Add CI using github Actions""

This reverts commit 84e0be02dd31d8d869cf789347b2b3d11ea1491a.

usuxnsnc committed:
Revert "Revert "Revert "Add CI using github Actions"""

This reverts commit 0a7dea4074099772631876648adca6a319566e6f.

bokuanT committed:
Merge pull request #3 from bokuanT/branch-BCD-Extension

Add remind feature, abstract LocalDateTime methods, improve code

usuxnsnc committed:
Merge branch 'master' into branch-A-Assertions

usuxnsnc committed:
Merge branch 'master' into branch-A-CodeQuality

usuxnsnc committed:
Fix minor code quality and check style errors

usuxnsnc committed:
Improve GUI and fix mimpr Ui bugs

Let's,
* Add colour, font, background colour and coloured textboxes
* Make profile pictures clipped to a circle
* Make Duke app window resizeable
* Add missing 'space' to unmark and mark index out of range error
message
* Remove wrongly formatted sad face in error messages

usuxnsnc committed:
Merge branch 'branch-A-BetterGui'

usuxnsnc committed:
Create user guide

Let's,
* Update README.md
* Make a basic user guide

usuxnsnc committed:
Fix minor bugs in Ui.png name and README

usuxnsnc committed:
Fix minor bugs in Ui.png naming and github visible readme

bokuanT committed:
Merge pull request #5 from bokuanT/branch-A-Assertions

Branch a assertions

bokuanT committed:
Merge pull request #6 from bokuanT/branch-A-CodeQuality

Branch a code quality
